Implement webgpu textureBindingViewDimension proposal to resolve texture view heuristics errors

alextechcc opened the pull request:
**Connections**
- Based on: https://github.com/teoxoy/wgpu/tree/texture_binding_view_dimension
- Based on: https://github.com/gpuweb/gpuweb/blob/main/proposals/compatibility-mode.md
- Works around #7428
- Currently is based on https://github.com/gfx-rs/wgpu/pull/8012 due to inability to run without it.

**Description**
- GLES backend currently spams the console with this error (https://github.com/gfx-rs/wgpu/blob/trunk/wgpu-hal/src/gles/mod.rs#L498) in bevy examples (https://github.com/bevyengine/bevy/issues/13115) and doesn't have a good way to use texture views as GLES 3 doesn't support texture views.

> Fundamental issue is that webgpu's textures and texture views are incompatible with opengl's textures and texture views because they are mappings of vulkan's images and image views.

 _Originally posted by @tasogare3710 in [#7428](https://github.com/gfx-rs/wgpu/issues/7428#issuecomment-2817209315)_

**Testing**
- Currently tested by replicating the error messages by creating a single layer D2Array and binding a texture view to it. Error is resolved and rendering is fixed. You can see this in the PR in examples/features/cube, I've disabled all examples except this one.
```bash
WGPU_BACKEND=gl cargo run --features gles --bin wgpu-examples cube
```

**Squash or Rebase?**
- Not ready to be merged yet, undecided.

**Checklist**
(keeping, as draft)
- [ ] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [ ] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [ ] Run `cargo xtask test` to run tests.
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

MarijnS95 committed:
Allow optional `RawDisplayHandle` in `InstanceDescriptor`

This allows platforms like GLES with EGL context backend to properly
initialize such that their `EGLDisplay` and `EGLContext` are compatible
with an incoming `Surface`, which is otherwise impossible if not for
hotpatching the context (and loosing all resources created on it) when a
surface comes in and that compositor connection becomes known.

This hotpatching is known to break quite a few systems and
initialization ordering (i.e. the alternative is forcing users to create
surfaces before querying adapters, which isn't feasible on platform like
Android), and is better avoided or at least made possible by using the
`(Raw)DisplayHandle` abstraction provided by `raw_display_handle` in the
way it's intended to be used (additionally, removing it from the surface
would be a sensible decision).

There may however be cases where users want to initialize multiple
`Instance`s or perhaps `Adapter`s for independent `RawDisplayHandle`s,
and certain backends may be designed specifically to cope with this
(i.e. Vulkan).

Note that this PR does nothing to address many other soundness problems
inside the EGL backend.

alextechcc committed:
DRAFT: Add textureBindingViewDimension compatibility mode to resolve GLES texture view incompatibility

alextechcc commented:
It's worth noting that this is currently a breaking change for every single create_texture call.

cwfitzgerald commented:
Sorry I'm not sure this got on my radar - besides being based on the other PR - would want the final version of the PR to be without it - we do want this!

alextechcc commented:
> Sorry I'm not sure this got on my radar - besides being based on the other PR - would want the final version of the PR to be without it - we do want this!

Sorry a bit confused at the wording here - do you mean rebasing without #8012 ? Can do if so - just difficult to test without it.

Is it okay that this is a breaking change for create_texture? If this is being used to implement WebGPU for Firefox I can imagine that the answer is that it can't be a feature flag.

cwfitzgerald commented:
Hey! We're really sorry this entirely fell through the cracks. We are interested in having this implemented, so if you're willing to move this to the latest trunk. If not feel free to close it - we'll be better about being responsive!

alextechcc commented:
On it!

alextechcc commented:
Okay, rebased on trunk, added a basic regression test, fixed the doctests/tests - but it's clear that there's a few shortcomings, and I'm not entirely sure if it's worth addressing this in a single PR.

I feel like the existing glsl heuristic doesn't match the compat spec (I believe it's a hack for particular use-cases?). A 6-layer D2 texture turns into "cube", but according to https://gpuweb.github.io/gpuweb/#dom-gpudevice-createtexture it looks like it should be 2d-array:
> Otherwise, if `dimension` is:
>
> * `"1d"` → set `textureBindingViewDimension` to `"1d"`.
> * `"2d"`:
>
>   * If the array layer count is `1`, set it to `"2d"`.
>   * Otherwise, set it to `"2d-array"`.
> * `"3d"` → set `textureBindingViewDimension` to `"3d"`.

I suppose if we have this field - the users experiencing the issue have a much better way to handle it, so we should probably follow the spec precisely.

Another aspect of this is - I believe that to properly follow the specification, the inferred binding view dimension should be set at the HAL layer (not invisibly handled in glsl), so that users (js/wasm, rust) can query it. Also, I saw a recommendation somewhere for wgpu that we should prefer that the HAL layer handle things like the translations before it hits the backend, which this probably falls under, but it's specific to glsl. So the tension is:
- Setting the default texture binding view dimension in the wgpu HAL layer makes sense, but probably only in compatibility mode. That way users can query it. Perhaps with compatibility mode off, it shouldn't do anything with the texture binding view dimension parameter? But compatibility mode isn't here yet.
- If it's set at the HAL layer, other backends get this confusing parameter passed. I'm assuming backends should ignore it if they don't need it.

Perhaps this points to compatibility mode landing before this being a better option?

Other missing pieces:
1. Validation missing, from https://gpuweb.github.io/gpuweb/
> - If textureBindingViewDimension is "2d", depthOrArrayLayers must be 1.
> - If textureBindingViewDimension is "cube", depthOrArrayLayers must be 6.
> - textureBindingViewDimension must not be "cube-array".
2. More potential validation / gotchas hunting. Dimension != texture binding view dimension cases. I probably missed a few, like in `create_texture_binding`.
3. deno_wgpu / wasm binding implementation missing. (Should this PR try to get that working?)
4. Missing changelog.
5. Existing error message for when the bound view dimension doesn't match the heuristic should probably change based on if the user provided a texture binding view dimension or not.
6. Some unrelated tests are currently skipped so I can focus on the regression test.

cwfitzgerald commented:
> I feel like the existing glsl heuristic doesn't match the compat spec (I believe it's a hack for particular use-cases?). A 6-layer D2 texture turns into "cube", but according to [gpuweb.github.io/gpuweb#dom-gpudevice-createtexture](https://gpuweb.github.io/gpuweb/#dom-gpudevice-createtexture) it looks like it should be 2d-array:

Correct - that will need to be updated. The current state is a hack to make the GL backend work vaguely correctly in most situations that people will end up using it.

> I suppose if we have this field - the users experiencing the issue have a much better way to handle it, so we should probably follow the spec precisely.

Yup.

> Perhaps this points to compatibility mode landing before this being a better option?

The current state of wgpu is that we don't support "compatibility mode" per se, as specified by the spec. We do however, support "downlevel" adapters which is sort of a proto-compatibility mode that actually has a lower capability floor: We support GLES 3.0/WebGL2 adapters, compat mode requires GLES 3.1. We can and should support this before we support formal compat mode because it fixes real problems for our users today.

What I expect this to look like: this is an optional field on a texture creation in `wgpu` and `wgpu-core`. It is passed through directly to wgpu-hal to inform the behavior of the GLES backends. Other backends ignore that value. When that value is passed, `wgpu-core` validates that only the correct texture view types are implemented. This is where the validation you mentioned in 1 would be implemented. wgpu-hal is allowed to assume it is being called correctly (which is why other backends can just flat out ignore this field), and core is responsible for validating in the way we need to validate. 

> deno_wgpu / wasm binding implementation missing. (Should this PR try to get that working?)

deno can just pass None when it calls this and web we should just drop it. Our bindings do not currently support compat mode (and only chrome supports it at all), so this is never useful.

Hope that helps! Feel free to ask more questions, happy to help!

alextechcc commented:
@cwfitzgerald It really does feel like the reading of the spec says textureBindingViewDimension is automatically picked for you and exposed - so that the user is able to check what was automatically determined (in deno, or in Rust). That would imply that it gets set at the core layer where the 'wgpu translations' happen, and that other backends would simply ignore it. With how wgpu is used, would breaking the spec like this be a problem?

Compat mode really does feel like a 'mode'. It infers the automatic default limits, makes this field exist (in JS) and auto-set, not just sets of limits. Maybe something like:
1. Before compat mode: No visibility in deno/rust what default got picked for you. GLSL uses it.
2. After compat mode: To follow the spec, if compatability mode is turned on, the textureBindingViewDimension is automatically selected at the wgpu-core layer, and it is visible to deno/rust. If compat mode is off. I don't know? This is where it breaks down.

cwfitzgerald commented:
> That would imply that it gets set at the core layer where the 'wgpu translations' happen, and that other backends would simply ignore it. With how wgpu is used, would breaking the spec like this be a problem?

Yeah we can break from the spec where needed as long as firefox has the tools needed to implement what they need to implement, and they aren't planning on implementing compat mode any time soon. Notably "compat" mode is something that you opt into in WebGPU where downlevel devices are not opt-in in wgpu. In WebGPU compat mode they can choose a value for the textureBindingViewDimension and then restrict all views to that dimension, because the user opted into that behavior. In order to make graceful fallback work, we should only require (or default) textureBindingViewDimension if a downlevel flag (say ARBITRARY_BINDING_VIEW_DIMENSION) is _not_ set. That way if the program is running on a API that allows arbitrary view dimensions, the user isn't forced to specify one at texture creation time. 

> No visibility in deno/rust what default got picked for you

We can still have visibility into what the current state is, it just needs to be return an option.

alextechcc commented:
Is non-strict deno on GLES considered supported? If you pick GLES with DENO_WEBGPU_STRICT_COMPLIANCE=0, before you could get away with the hacky fix for cube maps, but now you're stuck with the defaults.

I could add in the textureBindingViewDimension field to GPUTextureDescriptor and forward and handle it.
